Speed improvement to np.random.choice under replace=False and p=None

ryanpeach opened the pull request:
This is a fringe but also pretty common use-case for me. I want to choose a small sample of a large array, without replacement. The way this is currently written is O(n) where n is your array size. In the small case, however, with large arrays, this is very slow. The following algorithm's best performance is O(k) where k is your sample size. The worst case is unfortunately unbounded. There is always a probability that the random sample will continue to select the same values every time.

However, the probability that it will work in O(k) time, is the following:

![$P(O(k)) = \frac{n-1}{n} \cdot \frac{n-2}{n} \cdot ... \cdot \frac{n-k}{n} = \frac{n!}{n^k}$](https://latex.codecogs.com/gif.latex?P%28O%28k%29%29%20%3D%20%5Cfrac%7Bn-1%7D%7Bn%7D%20%5Ccdot%20%5Cfrac%7Bn-2%7D%7Bn%7D%20%5Ccdot%20...%20%5Ccdot%20%5Cfrac%7Bn-k%7D%7Bn%7D%20%3D%20%5Cfrac%7Bn%21%7D%7Bn%5Ek%7D)

Lets say we want it to choose this task if the probability of O(k) time is > .99 percent, and call that z.

![$P(O(k)) > .99 = z$](https://latex.codecogs.com/gif.latex?P%28O%28k%29%29%20%3E%20.99%20%3D%20z)

That probability will be such that n! < z*n^k where z = .99.

![$n! < z*n^k$](https://latex.codecogs.com/gif.latex?n%21%20%3C%20z%20n%5Ek)

Let's cancel z out instead and say z=1 (approximately). Thus, this algorithm should run whenever:

![z \approx 1.](https://latex.codecogs.com/gif.latex?z%20%5Capprox%201.)

![$n! < n^k$](https://latex.codecogs.com/gif.latex?n%21%20%3C%20n%5Ek)

Now, I haven't tested this code implemented in numpy, and I know running math.factorial is going to break at large numbers. Any way to simplify this inequality further? Does anyone want to dispute my math?

ryanpeach committed:
Added "small" option to choice.

This is a fringe but also pretty common use-case for me. I want to choose a small sample of a large array, without replacement. The way this is currently written is O(n). In the small case, however, this algorithm's best performance is O(k) where k is your sample size. The worst case is unfortunately unbounded. There is always a probability that the random sample will continue to select the same value every time. The probability that it will work in O(k) time, however, is n/k * (n-1)/k * (n-2)/k * ... * (n-k)/n==(n!)/n^k. Lets say we want it to choose this task if the probability of O(k) time is > .99 percent. That probability will be such that n! < z*n^k where z = .99. Let's cancel z out instead and say z=1 (approximately). Thus, this algorithm should run whenever n! < n^k.

rkern commented:
This general approach has been discussed previously and is blocked by stream-backwards compatibility issues until we figure out a good API for people to select newer (or older) implementations of the these methods: https://github.com/numpy/numpy/issues/2764

To your question about `math.factorial()`, there is an implementation of `loggam()` (the natural logarithm of the gamma function) in `distributions.c` for similar purposes. It's currently declared static for local use, but certain could be exposed for use in `mtrand.pyx`.

ryanpeach commented:
I don't know how to prove this, but all my graphics show that n! < n^k is linearly true so long as n<k<0. Could be divergent at high values though. I'd like a proof, but I bet there's some easy solution like k<.5n or something. I just want the switching to be non-arbitrary.

ryanpeach commented:
I have left my complete math and code snippet on the linked issue #2764

Thanks for your help.

ryanpeach committed:
Update mtrand.pyx

Fixed my math with better conditional proven in my pull request.

ryanpeach commented:
Rest of my proof here:

![$log_n(n!)<log_n(n^k)$](https://latex.codecogs.com/gif.latex?log_n%28n%21%29%3Clog_n%28n%5Ek%29)

![$log_n(n!)<k$](https://latex.codecogs.com/gif.latex?log_n%28n%21%29%3Ck)

![$log(\Gamma (1 + n))/log(n)<k$](https://latex.codecogs.com/gif.latex?log%28%5CGamma%20%281%20&plus;%20n%29%29/log%28n%29%3Ck)

[Sterlings approximation](https://www.johndcook.com/blog/2010/08/16/how-to-compute-log-factorial/) says:

![$log(\Gamma (1 + n)) = ((1+n)-1/2) \cdot log(1+n) - (1+n) + (1/2) \cdot log(2\pi)$](https://latex.codecogs.com/gif.latex?log%28%5CGamma%20%281%20&plus;%20n%29%29%20%3D%20%28%281&plus;n%29-1/2%29%20%5Ccdot%20log%281&plus;n%29%20-%20%281&plus;n%29%20&plus;%20%281/2%29%20%5Ccdot%20log%282%5Cpi%29)

![$= (n+1/2) \cdot log(1+n)- n + (1/2) \cdot log(2\pi) - 1$](https://latex.codecogs.com/gif.latex?log%28%5CGamma%20%281%20&plus;%20n%29%29%20%3D%20%28n&plus;1/2%29%20%5Ccdot%20log%281&plus;n%29-%20n%20&plus;%20%281/2%29%20%5Ccdot%20log%282%5Cpi%29%20-%201)

So to wrap it all up, our new code should run whenever:

![$\frac{(n+1/2) \cdot log(1+n)- n + (1/2) \cdot log(2\pi) - 1}{log(n)}<k$](https://latex.codecogs.com/gif.latex?%5Cfrac%7B%28n&plus;1/2%29%20%5Ccdot%20log%281&plus;n%29-%20n%20&plus;%20%281/2%29%20%5Ccdot%20log%282%5Cpi%29%20-%201%7D%7Blog%28n%29%7D%3Ck)

Implemented in code.

ryanpeach commented:
Sorry, just realized the inequality flipped.

ryanpeach committed:
Update mtrand.pyx

Missed a flipped inequality

ryanpeach committed:
Update mtrand.pyx

Fixed some basic bugs.

ryanpeach commented:
The unittest which fails due to a mismatch in arrays clearly is working, just isn't what's expected. I call shuffle so the in-order sorting of a length 3 array is just a coincidence most likely.

eric-wieser commented:
> The unittest which fails due to a mismatch in arrays clearly is working, just isn't what's expected

This is exactly the "stream-backwards compatibility" issue that @rkern is talking about. You function may be correct, but given the same `random.seed`, it produces a different shuffle order.

This is only fine if we allow users to request the original shuffle order somehow.

ryanpeach commented:
So we are going to ignore an obvious failing in the code because performing a library update might break your random seed? I think that's an update risk. I understand the slowness of pythonic sets being an issue, but I would have serious reservations about seed backward compatibility standing in the way of a better numpy.

ryanpeach committed:
Added @bashtage's variation for significant speed boost.

Added @bashtage 's variation on my algorithm using windowed randints of reducing size. Should give a significant speed boost.

ryanpeach committed:
Added legacy option at @eric-wieser's request.

ryanpeach commented:
Ok, so I've added a legacy option at @eric-wieser 's request and done a speed improving tweak at @bashtage 's suggestion. Should we add a depreciation warning as well so we can remove the legacy option in future versions?

ryanpeach committed:
Fixed the math and expanded a constant.

Found out I hadn't added the denominator to the equation. I also approximated a constant for better speed.

ryanpeach commented:
BTW, it seems that this condition is true, in general, for ratio's of (n, k) around {(10,6.3), (100,78.8), (1000, 855.8)}. That's pretty good! But we may want to timeit at these high values.

ryanpeach committed:
Restricting to pop_size>3

Below this number, the formula explodes to infinity or is less than 1.

ryanpeach committed:
Using new self.version instead.

Not sure if the version being greater than 0 is the right number to make, but I think this is all that's needed.

ryanpeach commented:
Now using self.version as per #5911

ryanpeach committed:
Fixed math, used loggam

Fixed some math, made cleaner by using loggam. Not sure if loggam is included however?

ryanpeach commented:
I just discovered my math was quite a bit off. P(O(k)) actually equals n!/(n^k*(n-k-1)!)  To save us all some time, the derivation comes out to log_n(n!)>k+log_n((n-k-1)!). Using the loggam function makes this cleaner subracting 1 from each and dividing by the base.

ryanpeach committed:
Update mtrand.pyx

ryanpeach commented:
That should be much better, but it may not compile if loggam isn't imported or the numbers are mistyped.

rkern commented:
Your vote with respect to stream compatibility is noted, but we get many more complaints when we break the stream. People rely on their seeded simulations reproducing the exact same results. Our longstanding (and thoroughly-discussed) policy for `numpy.random` is that we will break the stream if there is a correctness bug that needs to be fixed, but not in the case of performance improvements.

https://docs.scipy.org/doc/numpy/reference/generated/numpy.random.RandomState.html

Note that if you want to use `self.version`, then you need to base your PR off of the the branch in #5911, not `master`.

ryanpeach commented:
Ok, I'll rebase the pull request once I work everything out in terms of this code on my local machine.

bashtage commented:
I wouldn't rebase on that.  Unfortunately there isn't a clear path right now to multiple versions.  The correct method (IMO) isn't really possible since it isn't possible to override __new__ in Cython classes which could be is to swap classes so that nested classes.could be used. This method isn't perfect, and in particular it may not even work for set_state.  The other alternatives so far may get very messy over time.  

You may want to finish the PR and then just leave it open so that if the version problem is fixed then this can be incorporated.

ryanpeach commented:
Ok, well I've got it working on my local machine, but not with self (I'm importing numpy). I'll push it later.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: Kevin Sheppard <notifications@github.com>
Sent: Friday, October 13, 2017 4:02:50 AM
To: numpy/numpy
Cc: Ryan Peach; Author
Subject: Re: [numpy/numpy] Speed improvement to np.random.choice under replace=False and p=None (#9855)


I wouldn't rebase on that. Unfortunately there isn't a clear path right now to multiple versions. The correct method (IMO) isn't really possible since it isn't possible to override new in Cython classes which could be is to swap classes so that nested classes.could be used. This method isn't perfect, and in particular it may not even work for set_state. The other alternatives so far may get very messy over time.

You may want to finish the PR and then just leave it open so that if the version problem is fixed then this can be incorporated.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/numpy/numpy/pull/9855#issuecomment-336380436>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AOJryZR1nj2fEkf2TPjBQEp9E6v-nyTaks5srxkqgaJpZM4P2XFl>.

xoviat commented:
Probably the best method for backward compatibility is to move RandomState to a class factory function that returns instances of the versioned class:

```
def RandomState(seed=None, version=Latest):
    if version == 1.0:
        return RandomStateV1(seed=seed)
    elif version == 2.0:
        return RandomStateV2(seed=seed)
    else:
        return RandomStateV2(seed=seed)
```

And then you can just inherit the classes:

```
class RandomStateV1():
     pass

class RandomStateV2(RandomStateV1):
     pass

class RandomStateV3(RandomStateV2):
     pass
```

bashtage commented:
I don't think it is this simple since a multiversion RandomState will need pickle support as well as will need to accept the state from another version `get_state` and then behave like the requested version. 

One could use composition to achieve this effect, although it may be pretty messy.

xoviat commented:
> well as will need to accept the state from another version get_state and then behave like the requested version.

Why? If people are using a newer version then there's no use in preserving the old state.

rkern commented:
The point of versioning `RandomState` is to make sure that those using a newer version of `numpy` get the same results that they got when using an older version of `numpy`.

xoviat commented:
I understand that. Currently, when people need to use `RandomState`, they call the class initializer. What I am proposing is that the class initializer be changed to a function that dispatches to each particular version initializer. If there are class methods, then they would need to be changed to go through the factory function.

xoviat commented:
Each new version would in fact be an entirely new class that inherits from the previous version.

xoviat commented:
Idea: #9879.

rkern commented:
@bashtage The upgrade path wouldn't be so bad, actually. Current `RandomState` instances pickles reference `numpy.random.__RandomState_ctor` to construct the instance, not `numpy.random.RandomState()`. New classes could refer to other constructor functions, and `numpy.random.RandomState()` is free to do whatever, at least as far as unpickling old data is concerned.

bashtage commented:
Direct pickling/unpickling isn't the problem.  I was thinking about what would happen:

```
old_state = pickle.loads(old_state_str)
np.random.set_state(old_state)
identical_randoms_to_old_version = np.random.favorite_dist(size=1e6)
```

To correctly handle this the instance that is homed at `np.random` would need to be able to swap itself for an older version.  The only way I've been able to think about for this would be some form of composition. 

This is clearly bad behavior, but with the current guarantee it seems that the above command would be expected to work.  `get_state` already has logic for different state sizes since it was increased at some point in the past.

xoviat commented:
`np.random.set_state(old_state)` should raise an exception if the state from a previous version is loaded. This is much more clear and explicit than trying to hide this from the user and then them being surprised. The version information can be included in the pickling.

Also, the old version should be the default for now, but we should make people specify the version by a deprecation process.

xoviat commented:
Also #5911 is completely the wrong approach. I know people put hard work into that, but trying to force a version attribute to define class behavior is just bad.

bashtage commented:
The right approach is likely moving away from RandomState in a more
substantial way so that with modern implementations of random generators
can be added but also and more importantly modern core RNGs can be added -
ones that provide features like multiple streams or very fast jumps, such
as PCG,  xoroshift or the MLFG class.

On Tue, Oct 17, 2017, 21:41 xoviat <notifications@github.com> wrote:

> Also #5911 <https://github.com/numpy/numpy/pull/5911> is completely the
> wrong approach. I know people put hard work into that, but trying to force
> a version attribute to define class behavior is just bad.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/9855#issuecomment-337363995>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFU5RQiTzFIwuEcOoKLjdY_UhHpeUjZ-ks5stRDlgaJpZM4P2XFl>
> .
>

xoviat commented:
> The right approach is likely moving away from RandomState in a more
substantial way so that with modern implementations of random generators
can be added but also and more importantly modern core RNGs can be added -
ones that provide features like multiple streams or very fast jumps, such
as PCG,  xoroshift or the MLFG class.

Yes, and the class factory idea makes this possible. If you want to write a new random generator, just inherit from your favorite existing random generator or `RandomStateV1` and then include the type and version of yourself in the pickling information. Raise an exception if you get passed state information that doesn't match your type and version.

bashtage commented:
> Yes, and the class factory idea makes this possible. If you want to write a new random generator, just inherit from your favorite existing random generator or RandomStateV1 and then include the type and version of yourself in the pickling information. Raise an exception if you get passed state information that doesn't match your type and version.

I don't think making things possible is enough.  A big fix needs to be able to have a good API.  Raising every major version when someone wants to use a state from 1.N-1 doesn't seem like a good API.

bashtage commented:
WIthin the current RandomState structure I think it might make sense to use a dictionary to hold versioned functions.  This would avoid ending up on `RandomState27` since there would be only one RandomState (a bit like how `randint` handles 9 types).  In pure python it would be really easy to use a decorator to route for functions that had multiple versions.  I'm not sure if this would be a viable approach for an extension type.

xoviat commented:
They're completely different random generators though; they should have completely different classes. Swapping yourself out for a different class when you get passed an argument is just odd.

> Raising every major version when someone wants to use a state from 1.N-1 doesn't seem like a good API.

This is in fact the best possible API: 

```python
generator = RandomStateV1()
generator.beta()  # or whatever
```

It allows progress. It is explicit. People get the generator they want, and they know it won't change. But we can add a `RandomState()` function for backwards compatibility.

xoviat commented:
> This would avoid ending up on `RandomState27` since there would be only one RandomState

Also, do you realize what this means? It means that there will something like this:

```
if self.version == 1:
   pass
elif self.version == 2:
   pass
elif self.version == 3:
..
elif self.version == 27:
..
```

That is horrible for maintainability.

bashtage commented:
If you want to see the size cost of multiple `RandomStates`, see the file sizes here: https://pypi.org/project/randomstate/#description

There are 10 `RandomStates`, one for each different core PRNG, and the library is nearly the size on NumPy (bigger in some cases).

xoviat commented:
> WIthin the current RandomState structure I think it might make sense to use a dictionary to hold versioned functions. 

Actually, I read this again, and it's not a bad idea. But what would the public API look like?

ryanpeach commented:
Glad you all have picked this up. I think my math is right, I've been using it for a while now locally. I just don't know where to import loggam from. The versioning can be left up to you, but this is the right math:

(gammaln(n+1)-gammaln(n-k))/np.log(n) > k

ryanpeach commented:
![P(O(k))=\frac{n-1}{n}\cdot\frac{n-2}{n}\cdot...\cdot\frac{n-k}{n}=\frac{n!}{n^k(n-k-1)!}](https://latex.codecogs.com/gif.latex?P%28O%28k%29%29%3D%5Cfrac%7Bn-1%7D%7Bn%7D%5Ccdot%5Cfrac%7Bn-2%7D%7Bn%7D%5Ccdot...%5Ccdot%5Cfrac%7Bn-k%7D%7Bn%7D%3D%5Cfrac%7Bn%21%7D%7Bn%5Ek%28n-k-1%29%21%7D)

![P(O(k))=\frac{n!}{n^k(n-k-1)!}>z](https://latex.codecogs.com/gif.latex?P%28O%28k%29%29%3D%5Cfrac%7Bn%21%7D%7Bn%5Ek%28n-k-1%29%21%7D%3Ez)

![z\approx 1](https://latex.codecogs.com/gif.latex?z%5Capprox%201)

![\frac{n!}{n^k(n-k-1)!}>1](https://latex.codecogs.com/gif.latex?%5Cfrac%7Bn%21%7D%7Bn%5Ek%28n-k-1%29%21%7D%3E1)

![n!>n^k(n-k-1)!}>n^k](https://latex.codecogs.com/gif.latex?n%21%3En%5Ek%28n-k-1%29%21)

![log_n(n!)<log_n(n^k(n-k-1)!)](https://latex.codecogs.com/gif.latex?log_n%28n%21%29%3Clog_n%28n%5Ek%28n-k-1%29%21%29)

![log_n(n!)<log_n(n^k)+log_n((n-k-1)!)](https://latex.codecogs.com/gif.latex?log_n%28n%21%29%3Clog_n%28n%5Ek%29&plus;log_n%28%28n-k-1%29%21%29)

![log_n(n!)-log_n((n-k-1)!)<k](https://latex.codecogs.com/gif.latex?log_n%28n%21%29-log_n%28%28n-k-1%29%21%29%3Ck)

## Definitions:

![log_n(x)=\frac{ln(x)}{ln(n)}](https://latex.codecogs.com/gif.latex?log_n%28x%29%3D%5Cfrac%7Bln%28x%29%7D%7Bln%28n%29%7D)

![\Gamma(x+1)=x!](https://latex.codecogs.com/gif.latex?%5CGamma%28x&plus;1%29%3Dx%21)

Back to the proof:

![log_n(\Gamma(n+1))-log_n(\Gamma(n-k))<k](https://latex.codecogs.com/gif.latex?log_n%28%5CGamma%28n&plus;1%29%29-log_n%28%5CGamma%28n-k%29%29%3Ck)

![\frac{ln(\Gamma(n+1))}{ln(n)}-\frac{ln(\Gamma(n-k))}{ln(n)}<k](https://latex.codecogs.com/gif.latex?%5Cfrac%7Bln%28%5CGamma%28n&plus;1%29%29%7D%7Bln%28n%29%7D-%5Cfrac%7Bln%28%5CGamma%28n-k%29%29%7D%7Bln%28n%29%7D%3Ck)

Finally:

![\frac{ln(\Gamma(n+1))-ln(\Gamma(n-k))}{ln(n)}<k](https://latex.codecogs.com/gif.latex?%5Cfrac%7Bln%28%5CGamma%28n&plus;1%29%29-ln%28%5CGamma%28n-k%29%29%7D%7Bln%28n%29%7D%3Ck)

ryanpeach commented:
I’ll admit I must not understand the math, because the z can cancel at 1.0, but how could it really given that there’s always the possibility of repeating selection at k>1. Something is wrong.

ryanpeach committed:
Added the confidence (.99) and +1 term.

These make the selection tuneable by some confidence value in the probability that the algorithm will run in O(k) time.

Eight1911 commented:
At this point, do we have a consensus on what the new API should be? I'd be willing to take up the work if we have a finalized structure for the API.

rkern commented:
No, there isn't. We're discussing the subject on the mailing list right now.

https://mail.python.org/pipermail/numpy-discussion/2018-January/077608.html

ryanpeach committed:
Replaced .99 confidence with .5

If it's a greater than 50% chance of being a better algorithm, then it's worth trying right, because the expected value is better.

ryanpeach commented:
Could we maybe just make a flag for this. Make it default false?

mattip commented:
This PR should be part of the random number improvement [NEP 19](http://www.numpy.org/neps/nep-0019-rng-policy.html), specifically it should be part of the [bashtage/randomgen](https://github.com/bashtage/randomgen) project which will eventually be vendored or copied into NumPy.

bashtage commented:
I think the algorithm in #13163 is strictly faster than this method and always completed in O(size) ops

xref #13164

ryanpeach commented:
I do as well. Please disregard this PR. It was made in an attempt to provide a switchable algorithm, which could handle past-version compatibility. As that is being removed, I don't think it's useful.
